daemon/policy.go

```diff
@@ -206,27 +207,56 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
 	}
 
 	d.policy.Mutex.Lock()
+	// removedPrefixes tracks prefixes that we replace in the rules. It is used
+	// after we release the policy repository lock.
+	var removedPrefixes []*net.IPNet
 	if opts != nil {
 		if opts.Replace {
 			for _, r := range rules {
-				tmp := d.policy.SearchRLocked(r.Labels)
-				if len(tmp) > 0 {
-					d.dnsPoller.StopPollForDNSName(tmp)
+				oldRules := d.policy.SearchRLocked(r.Labels)
+				removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
+				if len(oldRules) > 0 {
+					d.dnsPoller.StopPollForDNSName(oldRules)
 					d.policy.DeleteByLabelsLocked(r.Labels)
 				}
 			}
 		}
 		if len(opts.ReplaceWithLabels) > 0 {
-			tmp := d.policy.SearchRLocked(opts.ReplaceWithLabels)
-			if len(tmp) > 0 {
-				d.dnsPoller.StopPollForDNSName(tmp)
+			oldRules := d.policy.SearchRLocked(opts.ReplaceWithLabels)
+			removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
+			if len(oldRules) > 0 {
+				d.dnsPoller.StopPollForDNSName(oldRules)
 				d.policy.DeleteByLabelsLocked(opts.ReplaceWithLabels)
 			}
 		}
 	}
 	rev := d.policy.AddListLocked(rules)
 	d.policy.Mutex.Unlock()
 
+	// remove prefixes of replaced rules above. This potentially blocks on the
+	// kvstore and should happen without holding the policy lock. Refcounts have
+	// been incremented above, so any decrements here will be no-ops for CIDRs
+	// that are re-added, and will trigger deletions for those that are no longer
+	// used.
+	//
+	if len(removedPrefixes) > 0 {
+		log.WithField("prefixes", removedPrefixes).Debug("Decrementing replaced CIDR refcounts when adding rules")
+
+		_ = d.prefixLengths.Delete(removedPrefixes)
+		if err := ipcache.ReleaseCIDRs(removedPrefixes); err != nil {
+			_ = d.prefixLengths.Delete(prefixes)
+			metrics.PolicyImportErrors.Inc()
+			log.WithError(err).WithField("prefixes", prefixes).Warn(
+				"Failed to release CIDRs in replaced rules during policy add with replace")
+
+			if errReleaseCIDR := ipcache.ReleaseCIDRs(prefixes); errReleaseCIDR != nil {
+				log.WithError(errReleaseCIDR).WithField("prefixes", prefixes).Warn(
+					"Failed to release CIDRs allocated for new rules during policy add with replace error cleanup")
+			}
+			return d.policy.GetRevision(), err
```

Comment:
This is a good point. Even if a re-insert happened because of the error, we have already leaked the release. I've made this a log that doesn't return, and matched the operation order to PolicyDelete.
Nothing will crash immediately but it will leak. I'm not sure how else to do this without refactoring all releases into an async scheme (as tgraf is implying). That would mean changing PolicyDelete and the ipcache implementation, at least, and perhaps any other places we want to release identities etc.